### ISSUE TITLE
Root implicit AD: freezing U*, Vh* in F drops a real term (2.5% gradient error, #715 Phase 1)

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#718]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

Sub-issue of #715, blocking PR #717.

## Summary

Root implicit AD for asymmetric CTMRG (arXiv:2607.15030 §V) treats the fixed-point decomposition tensors `U*`, `U_⊥`, `Vh*`, `Vh_⊥`, `s_star_inv` as **constants** inside the characteristic equations `F`. They are not — they depend on `p`. The implicit derivative therefore computes

```
dE/dp = ∂_p e − F̆ ∂_p F
```

and silently drops

```
− F̆ · (∂_c F) · (dc/dp)
```

That dropped term is worth a **2.5% gradient error**, and it is the only thing standing between `_ctm_root_implicit_asym.py` and a correct gradient.

The paper's licence for freezing them is that the `u` variable absorbs the change. That argument is about the *solution* `y*(p)`; it is not automatically true of the partial derivative, and it is measurably false in this implementation.

## Evidence

Directional derivative on a converged `D=2`, `χ=4` state, against explicit backprop (which itself matches a symmetric finite difference to ten digits):

```
implicit   = -0.810241844341
reference  = -0.817381274942
```

`dE/dc = −F̆ ∂_cF` measured directly. The energy has no explicit dependence on these constants, so this must vanish if freezing is valid:

| | \|dE/dc\| | kept-rotation part | subspace-tilt part |
|---|---|---|---|
| `U*[0]` | 1.455e-02 | 1.452e-02 | 9.36e-04 |
| `U*[1]` | 2.380e-03 | 2.287e-03 | 6.59e-04 |
| `U*[2]` | 2.576e-03 | 2.562e-03 | 2.65e-04 |
| `U*[3]` | 1.287e-02 | 1.287e-02 | 1.11e-04 |
| `Vh*[3]` | 1.736e-02 | 1.736e-02 | 1.60e-04 |

Nonzero, the right order of magnitude for the missing `7.14e-03`, and **dominated 15–100× by the kept-subspace rotation** rather than the subspace tilt.

## Everything else is ruled out

Each of these was measured, not argued:

| component | evidence |
|---|---|
| characteristic equations + root | `‖F(y*)‖ = 2.5e-16` |
| `∂_yF` non-singular | cond 7.9e3; after equilibration cond 3.6e2, zero null modes |
| environment + energy | `FD(E_implicit)` matches the reference to **8 digits** — `E_implicit(p)` is the correct function of `p` |
| `y → x` map | reconstructed energy `0.18833050074260` vs `0.18833050074260632` — 12 digits |
| linear solve | dense `np.linalg.solve` agrees with GMRES to **11 digits**; residual 2.3e-15 |
| choice of variables / scaling | equilibrated solve on a reparametrised system returns the identical gradient to 11 digits, as invariance requires |
| direct term `∂_p e` | plain VJP of the energy, correct by construction |

The entire error sits in the indirect term `−F̆ ∂_pF`: off by `0.0072` out of `0.4646`, i.e. 1.55% of that term.

## Why it is hard

Two routes, both currently blocked.

**1. Make `F` genuinely covariant**, so that freezing becomes legitimate. Six arrangements tried and measured, all recorded in `docs/plans/2026-07-29-715-phase1-modified-variables.md`:

| arrangement | error |
|---|---|
| baseline, diagonal `S` | 1.2e0 |
| **general-matrix `S` + matrix `S^{-1/2}`** | **1.1e-2 … 2.5e-2** ← current |
| quartic roots `(SS†)^{-1/4}`, `(S†S)^{-1/4}` in projectors | 2.0e-1 |
| modified variables `C̃`, `Ẽ` alone | no change (correct — a reparametrisation cannot move `dE/dp`) |
| quartic roots on the cut legs of `M` | 6.2e-2 / 9.4e-1 |
| combined covariance package | 2.0e-1 … 2.9e-1 |
| unsplit `S^{-1}` in the `y → x` map | refuted — energy 0.2473 vs 0.1883 |

The covariance argument made one correct prediction (general-matrix `S`, which fixed 120% → 2.5%) and then several wrong ones, so it is flawed somewhere not yet located.

**2. Compute `dU*/dp` and add the term.** The *tilt* component is well conditioned — governed by the truncation gap, the same quantity that makes Eq. 28 stable. But it is only 1/15 to 1/100 of the sensitivity. The dominant *rotation* component is exactly the `1/(s_i² − s_j²)` derivative that root implicit differentiation exists to avoid, so computing it directly would reintroduce the instability the method is for.

That tension — the freezing is invalid, but the dominant correction is the divergent quantity — is the substance of this issue. It is plausible the resolution lives in the parts of §V.3 that exist only as diagrams in the paper.

## What would resolve this

Any one of:

1. A formulation of `F` in which `dE/dU*` measurably collapses toward zero. The measurement above is a fast direct oracle — no need to infer from the end-to-end gradient.
2. A stable expression for the kept-rotation part of `dU*/dp`, or an argument that its contribution cancels against something.
3. Confirmation from the paper's figures (Eq. 65, Eqs. 78–80, Eq. 82) that the intended construction differs from what was reconstructed here.

## Methodological note for whoever picks this up

**`‖F(y*)‖` alone is not a sufficient oracle.** `F` will sit at machine precision on a nonphysical environment — that happened here and produced a wrong conclusion. Always pair the root residual with the energy through the `y → x` map.

Two other traps already hit and recorded:

- The tilde tensors come out at norm ~2e-3; eight of them underflow the reduced density matrix unless normalised. Without normalisation every reconstruction variant returns ~1e-13 and looks equally broken.
- Null-space counts must be taken **after** equilibration. Pre-equilibration the Jacobian had max singular value 1.4e-11, and a relative threshold produced a spurious "65 null modes ≈ 4χ²" reading that dissolved once scaled.

## Scope

Dense, 1×1 unit cell. #715 Phase 2 (unit cells) and Phase 3 (`SymmetricTensor`/fermionic) inherit whatever fix lands here, but neither is blocked on it structurally — their forwards and equations are independent.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

## Correcting the diagnosis in the issue body

I read the paper's §V.3 (arXiv:2607.15030v1 HTML source) rather than continuing to
reconstruct it. **The framing in the issue body is wrong** and should not be acted
on as written.

The body says the tension is intrinsic: "the freezing is invalid, but the dominant
correction is the divergent quantity." That is not the situation. Freezing `U*`,
`V*` is *legitimate* in the paper — Eq. 88 restricts the isometry variations to
their null spaces exactly so that it is. What makes Eq. 88 licensed is a gauge
covariance that this implementation never established.

### What §V.3 actually specifies

The covariance repair is a **package**, in two steps:

1. Remove the inverse square roots from the projector definitions (Eqs. 66/67)
   *and* from the renormalised corners and edges, putting the `s_α` explicitly on
   the **edges**. This yields the modified `C̃`, `Ẽ`, `P̃^L`, `P̃^R` of Eqs. 82,
   74, 75. Quoting: *"The appearance of inverse singular values {s_α} on the edges
   can be straightforwardly generalized to larger environments to systematically
   ensure gauge invariance."*
2. The Eq. 65 environment matrix is formed by cutting links **between edge
   tensors**, which in modified variables leaves a dangling `√s_α` on the cut outer
   legs — *"a non-unitary matrix transformation [that] consequently alters the SVD
   decomposition beyond a simple gauge transformation."* Replace those dangling
   roots on the left/right cuts with `s^L_α`, `s^R_α` of Eq. 73, whose transformation
   laws (Eq. 87) *are* covariant. This *"leads to the expressions for the infinite
   environment matrix on the left hand side of Eqs. 78 to 80."*

Only after both steps does Eq. 88 hold, and only then is freezing valid.

### Why the exoneration table was misleading

Every arrangement in that table was tried **in isolation**, and each failed in a way
that looked like a refutation:

| what I recorded | why it was not a refutation |
|---|---|
| `C̃`, `Ẽ` alone → "no change (correct — a reparametrisation cannot move dE/dp)" | True in isolation, and that is the trap. Without step 2 and a matching adjoint it *is* only a reparametrisation. It is not optional; it is half of the prerequisite. |
| quartic roots **in the projectors** → 2.0e-1 | Wrong location. §V.3 puts `s^L`/`s^R` on the **cut legs of the environment matrix** in Eqs. 78–80, not inside Eqs. 74/75. |
| quartic roots on the cut legs of `M` → 6.2e-2 / 9.4e-1 | Right location, but without step 1 there is no covariant edge to cut. |
| "combined covariance package" → 2.0e-1…2.9e-1 | Was not the paper's combination — it lacked the Eq. 82 adjoint below. |

### The concrete departure

`AsymRoot.env` holds the **regular** corners/edges straight from `sweep`, so Eq. 82's
map is the identity here, and correspondingly the code sets

```python
tuple(jnp.zeros_like(x) for x in root.s),   # S̆ = 0
```

The paper zeroes **only** `ŭ` and `v̆`, because `U` and `V` do not appear in the
energy. `S̆` is explicitly required to be nonzero: *"from the adjoints x̆ of the
corner and edge tensors we can obtain adjoints of the variables `C̃_α`, `Ẽ_α` and
`S̃_α` by explicitly backpropagating through the defining relation Eq. 82."*

Note `S̆ = 0` is *self-consistent* with the identity map, so this is not a one-line
bug — the implemented scheme is coherent, just unlicensed. Fixing it means
implementing the package, not patching the adjoint.

### One more requirement worth flagging

§V.3 is explicit that forward and reverse differ operationally: the forward keeps
`S_α` **real and diagonal**, while *"throughout the reverse mode we treat `S_α` as
complex non-diagonal matrices. This begins with defining the modified corners and
edges in Eq. 82."* Promoting `S` to a general matrix (already done, and what took
the error from 1.2e0 to 2.5e-2) is one part of that; it belongs with Eq. 82, not
on its own.

### Revised plan

Implement §V.3 as a unit: `s` onto the edges + modified `C̃`/`Ẽ`/`P̃^L`/`P̃^R`
(Eqs. 82, 74, 75), `s^L`/`s^R` on the cut legs in Eqs. 78–80, and `ў` built by
backpropagating `x̆` through Eq. 82 so `S̆ ≠ 0`.

The `dE/dU*` oracle in the issue body remains the right fast check: if the package
is correct it must collapse toward zero, and it does not require the end-to-end
gradient to measure.

Reference implementation exists — PEPSKit.jl / TensorKit.jl, cited as Ref. 46 in the
paper — and is worth diffing against before re-deriving anything further.
